src/storage-backend.js

```diff
@@ -0,0 +1,655 @@
+   * value 'undefined' will be returned.
+   */
+  async readAddressFromCache(rawUrl) {
+    assert(rawUrl);
+    let key = encodeURL(rawUrl);
+    let jsonVersion = await this.memcached.get(key);
+    debug(`${this.id} MEMCACHED GET: ${key}, ${jsonVersion}`);
+    return jsonVersion ? JSON.parse(jsonVersion) : undefined;
+  }
+
+  /**
+   * Request that a file be stored in the cache.  We use SQS as our messaging queue
+   * for these copy requests.  This operation can be done without the StorageBackend
+   * being in listening mode.  Listening mode is strictly for consumers of this queue
+   */
+  async requestPut(rawUrl) {
+    let sqsParams = {
+      QueueUrl: this.copyRequestQueue,
+      MessageBody: JSON.stringify({
+        url: rawUrl,
+      }),
+    };
+    await this.sqs.sendMessage(sqsParams).promise();
+  }
+
+  /** 
+   * Create a name for a given storage backend.  Since we can have a reasonable
+   * guess of an appropriate default action, we do have an implementation in
+   * the base class.  This implementation assumes that you'll essentially have
+   * a single parameter in a URL in order to resolve it in the backing storage.
+   * This is not the case in S3, as S3 does require a bucket name to be
+   * present.
+   *
+   * Overriding this is only needed if you require more than a single token
+   * to determine where a file is in a given backing store
+   */
+  backendAddress(rawUrl) {
+    let key = encodeURL(rawUrl);
+    assert(key.length <= 1024, 'key must be 1024 or fewer unicode characters');
+    return encodeURL(key);
+  }
+
+  /**
+   * Determine when the file in the backend expires
+   */
+  async expirationDate(headers) {
+    return this._expirationDate(headers);
+  }
+
+  /**
+   * Retreive the expiration time of object in S3
+   */
+  async _expirationDate(headers) {
+    throw new Error('Putting not yet implemented');
+  }
+
+  /**
+   * Stub method for the backend specific operations to upload the resource
+   * represented by 'inStream' to the address represented by 'name'.  The
+   * original URL, contentType, redirects and upstreamEtag will be passed into
+   * this method for use in the upload process.
+   *
+   * The backend implementation must return a promise that resolves to the
+   * information that can be used to identify the copy of this resource that it
+   * contains.
+   */
+  async _put(name, inStream, rawUrl, contentType, redirects, upstreamEtag) {
+    throw new Error('Putting not yet implemented');
+  }
+
+  // Create a real URL from a backend address
+  _backendAddressToUrl(backendAddress) {
+    // We throw here because it's impossible to know how the implementing class
+    // will map this backend address into a real url.  Simply decoding the URL
+    // would mean that the cache isn't used and that's the only thing we know
+    // how to do with this data 
+    throw new Error('Unimplemented');
+  }
+
+  async _expire(name) {
```

Comment:
Okay, I'm getting confused... We have following concepts:
 * backendAddress
 * backendAddressUrl
 * name
 * rawUrl

I think `name` and `backendAddress` are the same, right?


And if I get it right you invented `backendAddress` because how else could you implement force purge, right?
(Not that purge is not important to me, I don't think we can do it, deletes on S3 are eventually consistent, so what is the point vs. just letting the life-cycle do it's thing. As anything relying on purge will fail).

IMO, purge in S3 could rely on decoding the `backendAddressUrl`. Granted it would be messy.
So if we really want an object to represent a `backendAddress` which I don't like, let's make an object that has a property called `url`. With the rest of the properties being loosely typed, but stilled stored in cache.